packages/nitro-protocol/contracts/MultiAssetHolder.sol

```diff
@@ -429,8 +426,8 @@ contract MultiAssetHolder is IMultiAssetHolder, StatusManager {
                         // only if specified in supplied indices, or we if we are doing "all"
                         // reduce the new allocationItem.amount
                         newAllocation[i].amount -= affordsForDestination;
-                        // increase the relevant payout
-                        payouts[k] += affordsForDestination;
+                        // the amount subtracted should be paid out to this destination
+                        // increase the total payouts
                         totalPayouts += affordsForDestination;
                         // move on to the next supplied index
                         ++k;
```

Comment:
> The problem with the current implementation is that inside _computeNewAllocationWithGuarantee, k remains at its initial value of 0

I still don't understand how this could be true, as `k` is incremented here!

As such, I have a really hard time understanding the bug. If `k` is not incremented, will the new allocation be correctly computed?

Reply:
`k` is incremented here, *but only when the condition on L420 is true*. While `j=0`, the condition is never true in the example I gave in the PR description, so `k` will *not be incremented*. When `j=1, i=1` the condition is true (we found `I` from the guarantee in the target allocation), and `payouts[k]` is increased  -- but  `k` still has its initial value of `0`.

`k` is then incremented afterwards, but we have now run out of funds so will break out of the loops without any more payouts being changed.

When `indices=[]`, `payouts[k]` will be interpreted as the payout for `allocation[k].destination`. So in this case we have `payouts[0] = 10` coins for `allocation[0].destination = x`. Which is wrong. It is in this sense that "not incrementing `k` when we don't find a guarantee destination in the target" is a problem. Of course if we no longer have that interpretation for `payouts`, then it may no longer be a problem -- which is why the fixed implementation having the same logic for incrementing `k` is OK.

There is no such problem when computing the new allocation, since that doesn't use the index `k`; it uses the index `i`. 

If you are having a hard time understanding the bug, I think it is worth remembering that that is a secondary consideration. The more important things to understand are:
1. That the current implementation has a bug. You can see this from the new test cases, which fail before the fix is applied.
2. That the fixed implementation is correct. 

It might be that it is difficult to understand the bug because it is difficult to understand the current implementation. Happily, the fixed implementation is much simpler anyway -- which is probably reason enough to accept if there were no bug in the current implementation.